de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java

```diff
@@ -0,0 +1,146 @@
+package de.fu_berlin.inf.dpp.server.session;
+
+import java.util.HashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.log4j.Logger;
+import de.fu_berlin.inf.dpp.negotiation.SessionNegotiation;
+import de.fu_berlin.inf.dpp.negotiation.OutgoingSessionNegotiation;
+import de.fu_berlin.inf.dpp.negotiation.IncomingSessionNegotiation;
+import de.fu_berlin.inf.dpp.negotiation.ProjectNegotiation;
+import de.fu_berlin.inf.dpp.negotiation.AbstractOutgoingProjectNegotiation;
+import de.fu_berlin.inf.dpp.filesystem.IProject;
+import de.fu_berlin.inf.dpp.monitoring.NullProgressMonitor;
+import de.fu_berlin.inf.dpp.negotiation.AbstractIncomingProjectNegotiation;
+import de.fu_berlin.inf.dpp.negotiation.NegotiationTools;
+import de.fu_berlin.inf.dpp.server.ServerConfig;
+import de.fu_berlin.inf.dpp.server.progress.ConsoleProgressIndicator;
+import de.fu_berlin.inf.dpp.session.INegotiationHandler;
+import de.fu_berlin.inf.dpp.session.ISarosSessionManager;
+import de.fu_berlin.inf.dpp.util.NamedThreadFactory;
+
+public class NegotiationHandler implements INegotiationHandler {
+
+    private static final Logger log = Logger
+        .getLogger(NegotiationHandler.class);
+
+    private final ISarosSessionManager sessionManager;
+    private final ThreadPoolExecutor sessionExecutor = new ThreadPoolExecutor(0, 4, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(), new NamedThreadFactory(
+        "ServerSessionNegotiation-"));
+    private final ThreadPoolExecutor projectExecutor = new ThreadPoolExecutor(0, 10, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(), new NamedThreadFactory(
+        "ServerProjectNegotiation-"));
+
+
+    public NegotiationHandler(ISarosSessionManager sessionManager) {
+        sessionManager.setNegotiationHandler(this);
+        this.sessionManager = sessionManager;
+    }
+
+    @Override
+    public void handleOutgoingSessionNegotiation(
+        final OutgoingSessionNegotiation negotiation) {
+
+        sessionExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                SessionNegotiation.Status status;
+                if (ServerConfig.isInteractive()) {
+                    status = negotiation
+                        .start(new ConsoleProgressIndicator(System.out));
+                } else {
+                    status = negotiation
+                        .start(new NullProgressMonitor());
+                }
+
+                switch (status) {
+                    case OK:
+                        sessionManager.startSharingProjects(negotiation.getPeer());
+                        break;
+                    case ERROR:
+                        log.error("ERROR running session negotiation: "+negotiation.getErrorMessage());
```

Comment:
FYI if you write

String s = a + b + c 

This is only syntactic sugar and the compiler will transform this into

String s = new StringBuilder().append(a).append(b).append(c).toString();

where append() will invoke the toString() method on the object you are passing in

TLDR: it is not necessary to call the toString() method as it is done automatically.